### DISCUSSION TITLE
關於“通过SToken获取账号Auth Key B”返回-100 ERROR

### DISCUSSION THREAD

**Discussion by Vocaloid2048** [UIGF-org/mihoyo-api-collect#62] [Q&A]
想向各位大佬們請教一下，根據[UIGF文檔](https://uigf.org/zh/mihoyo-api-collection/hoyolab/user/token.html#%E9%80%9A%E8%BF%87stoken%E8%8E%B7%E5%8F%96%E8%B4%A6%E5%8F%B7auth-key-b)，使用這個API需要提供：
- header : 
 - ds : 使用LK2 的 DS1生成結果 e.g. (1709627103,nA3iZH,0b8614836a28b996225922f95c5fa130) 
 - x-rpc-app_version : salt的對應版本 (e.g. 2.68.1)
 - x-prc-client_type : 5 

### 但在cookies部分，我嘗試過放入剛剛從[getMultiTokenByLoginTicket](https://api-takumi.mihoyo.com/auth/api/getMultiTokenByLoginTicket) 生成，已經確認正確的SToken : 
![螢幕截圖 2024-03-05 17 17 45](https://github.com/UIGF-org/mihoyo-api-collect/assets/47070571/39dbbead-5350-4fe0-b08a-d2b28bc7c1fd)

調用genAuthKey的部分
![螢幕截圖 2024-03-05 17 19 00](https://github.com/UIGF-org/mihoyo-api-collect/assets/47070571/2e11f2fb-fb8c-486e-9f09-26f3a5039ce7)

![螢幕截圖 2024-03-05 17 13 10](https://github.com/UIGF-org/mihoyo-api-collect/assets/47070571/4d31f986-a891-4a15-8c7c-feb102fb3ca7)

測試過在cookies欄位內添加mid (取於ltuid_v2和ltmid_v2也試過，都是出現以上結果）
## 感謝

**Comment by BTMuli** [comment]
[GenAuthkey Test](https://github.com/BTMuli/TGAssistant/blob/master/test/genAuthkey.test.ts)

**Comment by Vocaloid2048** [comment]
> [GenAuthkey Test](https://github.com/BTMuli/TGAssistant/blob/master/test/genAuthkey.test.ts)

十分感謝幫忙！先慢慢研究一下~ thxxx

**Comment by Vocaloid2048** [comment]
> [GenAuthkey Test](https://github.com/BTMuli/TGAssistant/blob/master/test/genAuthkey.test.ts)

想再向大佬們請教一下，目前透過SToken v1 獲取了 SToken v2，SToken v2獲取了AuthKeyB（auth_appid = webview_gacha)
![螢幕截圖 2024-03-07 11 40 58](https://github.com/UIGF-org/mihoyo-api-collect/assets/47070571/bc552ad9-046c-4948-8517-6fc4466c79f3)

但如果我直接把AuthKeyB完封不動地貼上在query的部分，它卻反而說AuthKey無效
![螢幕截圖 2024-03-07 11 45 21](https://github.com/UIGF-org/mihoyo-api-collect/assets/47070571/10c34396-6b54-4e1a-b9ad-4ea213237e36)
也試過url encode,但也沒效...
先感謝各位大佬的幫忙

**Comment by BTMuli** [comment]
代码没什么问题，~~你把参数改成原神的就正常了~~

铁道获取祈愿记录目前采取的都是缓存或者代理，通过stoken生成的authkey不能用，抓包或者缓存获取的可以 #10

**Comment by Vocaloid2048** [comment]
> 代码没什么问题，~~你把参数改成原神的就正常了~~
> 
> 铁道获取祈愿记录目前采取的都是缓存或者代理，通过stoken生成的authkey不能用，抓包或者缓存获取的可以 #10

感謝🙇 ~~這就難辦了QQ~~

**Comment by Vocaloid2048** [comment]
> 代码没什么问题，~你把参数改成原神的就正常了~
> 
> 铁道获取祈愿记录目前采取的都是缓存或者代理，通过stoken生成的authkey不能用，抓包或者缓存获取的可以 #10

不好意思 又要拜託大佬了🙇🙇
剛才想了幾小時，既然鐵道祈願的authkey在每次開啟遊戲的時候都是不同的
那麼應該是有一道算法（但不確定是否放在App本體、伺服器還是動態的script）。
看了App本體，沒有相關的算法 (GetAuthKeyEntity.java)
抓包沒有看到有調用 ([https://sdk.hoyoverse.com/sw.js](https://sdk.hoyoverse.com/sw.js) 沒有關係吧？)
動態的script（不太清楚怎樣看.bytes的 Lua檔案）

在看抓包紀錄的時候 鐵道祈願的authkey是無緣無故地突然出現在query的部分
實在想不明白 期盼大佬有找到 除了透過抓包/讀取webCache緩存以外的解決方案
~~感覺這個要求真的太強人所難 真的不好意思🙇🙇~~
![image](https://github.com/UIGF-org/mihoyo-api-collect/assets/47070571/b3b48572-2f5e-4f42-b12f-2c64b0404755)
百分感謝幫忙🙇🙇

**Comment by K0max** [comment]
> > 代码没什么问题，~你把参数改成原神的就正常了~
> > 铁道获取祈愿记录目前采取的都是缓存或者代理，通过stoken生成的authkey不能用，抓包或者缓存获取的可以 #10
> 
> 不好意思 又要拜託大佬了🙇🙇 剛才想了幾小時，既然鐵道祈願的authkey在每次開啟遊戲的時候都是不同的 那麼應該是有一道算法（但不確定是否放在App本體、伺服器還是動態的script）。 看了App本體，沒有相關的算法 (GetAuthKeyEntity.java) 抓包沒有看到有調用 (https://sdk.hoyoverse.com/sw.js 沒有關係吧？) 動態的script（不太清楚怎樣看.bytes的 Lua檔案）
> 
> 在看抓包紀錄的時候 鐵道祈願的authkey是無緣無故地突然出現在query的部分 實在想不明白 期盼大佬有找到 除了透過抓包/讀取webCache緩存以外的解決方案 ~感覺這個要求真的太強人所難 真的不好意思🙇🙇~ ![image](https://private-user-images.githubusercontent.com/47070571/310937522-b3b48572-2f5e-4f42-b12f-2c64b0404755.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE3MDA3MzAsIm5iZiI6MTcxMTcwMDQzMCwicGF0aCI6Ii80NzA3MDU3MS8zMTA5Mzc1MjItYjNiNDg1NzItMmY1ZS00ZjQyLWIxMmYtMmM2NGIwNDA0NzU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI5VDA4MjAzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZlOTViYzNkYjk5ZmJkY2VkYjhmYjk0ODAyZGQ4NjIzNzJkY2QyN2FmMTNhMTZhYzVhMjQyYjdjMTkyOTEyODMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hDpRP3Hm-e1znYnzWDb82W_-nlK53sAqsk1znhDLYR0) 百分感謝幫忙🙇🙇

如果是莫名其妙出现的话，会不会 authkey 是在本地使用特定加密算法生成的？因为 #10 中提到 authkey 是固定的684位

**Comment by Vocaloid2048** [comment]
> > > 代码没什么问题，~你把参数改成原神的就正常了~
> > > 铁道获取祈愿记录目前采取的都是缓存或者代理，通过stoken生成的authkey不能用，抓包或者缓存获取的可以 #10
> > 
> > 
> > 不好意思 又要拜託大佬了🙇🙇 剛才想了幾小時，既然鐵道祈願的authkey在每次開啟遊戲的時候都是不同的 那麼應該是有一道算法（但不確定是否放在App本體、伺服器還是動態的script）。 看了App本體，沒有相關的算法 (GetAuthKeyEntity.java) 抓包沒有看到有調用 (https://sdk.hoyoverse.com/sw.js 沒有關係吧？) 動態的script（不太清楚怎樣看.bytes的 Lua檔案）
> > 在看抓包紀錄的時候 鐵道祈願的authkey是無緣無故地突然出現在query的部分 實在想不明白 期盼大佬有找到 除了透過抓包/讀取webCache緩存以外的解決方案 ~感覺這個要求真的太強人所難 真的不好意思🙇🙇~ ![image](https://private-user-images.githubusercontent.com/47070571/310937522-b3b48572-2f5e-4f42-b12f-2c64b0404755.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE3MDA3MzAsIm5iZiI6MTcxMTcwMDQzMCwicGF0aCI6Ii80NzA3MDU3MS8zMTA5Mzc1MjItYjNiNDg1NzItMmY1ZS00ZjQyLWIxMmYtMmM2NGIwNDA0NzU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI5VDA4MjAzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZlOTViYzNkYjk5ZmJkY2VkYjhmYjk0ODAyZGQ4NjIzNzJkY2QyN2FmMTNhMTZhYzVhMjQyYjdjMTkyOTEyODMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hDpRP3Hm-e1znYnzWDb82W_-nlK53sAqsk1znhDLYR0) 百分感謝幫忙🙇🙇
> 
> 如果是莫名其妙出现的话，会不会 authkey 是在本地使用特定加密算法生成的？因为 #10 中提到 authkey 是固定的684位

我猜應該是了 但比較尷尬的是 目前我們沒法找到那算法
如果是封裝在動態script内的話就更麻煩了

**Reply by ChenVAV** [reply]
铁道的祈愿链接似乎是解析服务器返回的数据[#61](https://github.com/UIGF-org/mihoyo-api-collect/discussions/61?sort=new#discussioncomment-13034869)直接拿到的。

**Reply by iceprosurface** [reply]
可以确定肯定不是 https://sdk.hoyoverse.com/sw.js 这个是前端用的 service worker，我反混淆了一下，里面只有 cache 相关的逻辑

**Reply by Lightczx** [reply]
和游戏业务相关的AuthKey均由游戏通过KCP向服务器请求生成，目前只有原神的祈愿/颂愿可以使用米游社的genAuthKey，其他的都需要游戏内的GenAuthKeyReq
